Move c10d to libtorch(_cuda)

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59563 Move c10d to libtorch(_cuda)**

Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

lw committed:
Move c10d to libtorch(_cuda)



Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -1210,9 +1239,30 @@ endif()
-  target_compile_definitions(torch_cpu PRIVATE
-    USE_DISTRIBUTED
-  )
+  target_include_directories(torch_cpu PUBLIC ${TORCH_SRC_DIR}/lib)
```
Can you please add a comment, why this is needed?

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -1210,9 +1239,30 @@ endif()
+  if(USE_NCCL AND USE_C10D_NCCL)
+    if(USE_ROCM)
+      target_compile_definitions(torch_hip PUBLIC USE_C10D_NCCL)
+    elseif(USE_CUDA)
```
Nit: `USE_NCCL` implies that either USE_CUDA or USE_ROCM is defined
```suggestion
    else()
```

malfet commented on tools/build_variables.bzl:
```diff
@@ -314,7 +314,26 @@ core_sources_full = core_sources_full_mobile + [
 libtorch_core_sources = sorted(core_sources_common + core_sources_full + core_trainer_sources)
 
-libtorch_distributed_sources = [
+libtorch_distributed_windows_sources = [
```
Nit: Please consider renaming `libtorch_distributed_windows_sources` to `libtorch_distribtued_base_sources`, as there are nothing windows specific about them
```suggestion
libtorch_distributed_base_sources = [
```

lw replied:
I called them `windows` because they're the only sources that are safe to build under Windows. I'll rename them to `base` as suggested, and add a comment to clarify why we need to separate them.

malfet commented on tools/build_variables.bzl:
```diff
@@ -314,7 +314,26 @@ core_sources_full = core_sources_full_mobile + [
+    "torch/lib/c10d/Utils.cpp",
+]
+
+libtorch_distributed_sources = libtorch_distributed_windows_sources + [
```
If filelists are split into base and c10d source if we replace `if();add();else();add();endif()` into `add(base);if();add(extra);endif();`
```suggestion
libtorch_distributed_c10d_sources =  [
```

lw replied:
I don't want to use `c10d` in the name because after this change we'll be able to also move the RPC layer from `libtorch_python` to `libtorch` and its files would be added to this list as well. But I'm fine with calling it `extra` as suggested.

malfet replied:
How about `libtorch_distributed_extra_sources`

lw replied:
Yes that's what I went with. I also left `libtorch_distributed_sources` defined as the union of base+extra, since it's handy internally. Hope this is fine?

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -653,6 +657,21 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
         ${TORCH_SRC_DIR}/csrc/cuda/nccl.cpp)
     endif()
+    if(USE_DISTRIBUTED)
+      if(WIN32)
+        if(BUILD_SPLIT_CUDA)
+          append_filelist("libtorch_cuda_distributed_windows_sources" Caffe2_GPU_SRCS_CPP)
+        else()
+          append_filelist("libtorch_cuda_distributed_windows_sources" Caffe2_GPU_SRCS)
+        endif()
+      else()
+        if(BUILD_SPLIT_CUDA)
+          append_filelist("libtorch_cuda_distributed_sources" Caffe2_GPU_SRCS_CPP)
+        else()
+          append_filelist("libtorch_cuda_distributed_sources" Caffe2_GPU_SRCS)
+        endif()
+      endif()
```
Two suggestions (probably made typo in the code):
 - define filelist target based on `BUILD_SPLIT_CUDA` option
 - Change additive `distributed` into standalone list and add it if not WIN32
```suggestion
        if(BUILD_SPLIT_CUDA)
          set(_target "Caffe2_GPU_SRCS_CPP)
        else()
          set(_target "Caffe2_GPU_SRCS")
        endif()
        append_filelist("libtorch_cuda_base_distributed_sources" ${_target_})
      if(NOT WIN32)
          append_filelist("libtorch_cuda_distributed_c10d_sources" ${_target})
      endif()
```

lw replied:
Done.

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]

lw committed:
Update on "Move c10d to libtorch(_cuda)"


Differential Revision: [D28932239](https://our.internmc.facebook.com/intern/diff/D28932239/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28932239/)!

[ghstack-poisoned]
